### PR TITLE
contracts: Add `salt` argument to contract instantiation

### PR DESCRIPTION
**Pull request by athei**
This PR adds a `salt: &[u8]` argument to the `instantiate` dispatchable and the `seal_instantiate` host function. The new formula for contract address derivation is:

```
hash(deploying_address ++ code_hash ++ salt)
```

This is akin to Ethereum's CREATE2 semantic.

**This PR can be reviewed on a commit by commit basis. The description of each commit contains more information on what happened there.**

## Motivation

Currently, the formula to to derive a new contract address looks like this:
```
hash(deploying_address ++ code_hash ++ hash(input_bytes))
```

This formula has two problems: First, a generic UI cannot help the user to deploy the same contract multiple times. Even though the formula theoretically allows for deploying the same contract from the same sender multiple times this support needs to be baked into the contract code and the UI needs to be aware of that. Today, this is not the case. Second, hashing the contract input for every input is wasteful because it could be a a large data blob while something shorter would suffice as input for the address derivation.

The change in this PR solves these problems: The input data is no longer hashed and a generic UI could suggest the account nonce or some off-chain generated random number as default value for the salt. Since the `salt` is just raw bytes it could also be used to implement the old behaviour by supplying the hash of the input data as `salt`.

Please note that the decision against implementing CREATE semantic where the account nonce is used for the contract address is intentional: The new salt semantic is strictly more powerful and is needed anyways for counterfactual contract deployment.

## Implementation Hurdles

Up until now the actual contract address derivation was performed by a runtime configuration supplied function. Theoretically, pallet_contracts wasn't able to make any assumptions about how the contract address was derived. In practice, assumptions must be made in order to write proper benchmarks which use the actual runtime configuration. Additionally, the whole construct is a workaround because `pallet_system::Trait::AccountId` is missing some trait bounds that allow proper `AccountId` derivation. This PR makes the pallet_contracts require those trait bounds and subsequently internalizes runtime passed derivation functions. The fallout from that is minimized by grouping many free standing functions to inherent functions to reduce the littering with `where` clauses. See [rfc-2089](https://github.com/rust-lang/rfcs/blob/master/text/2089-implied-bounds.md).

### Client
@jacogr 
Clients need to be aware of an additional argument to the `instantiate` dispatchable which has a new argument `salt`:
https://github.com/paritytech/substrate/blob/dc8b4ce7a5bb2fafd99d885f3e763b6bb5086774/frame/contracts/src/lib.rs#L513

### Contracts
@Robbepop @seanyoung
Contracts need to be aware that the `seal_instantiate` host function has the same new argument:
https://github.com/paritytech/substrate/blob/1809417f3dbf5276c7b5a437ddd2b94314e7c771/frame/contracts/src/wasm/runtime.rs#L860-L861

### Runtime Integration
@ascjones 
`DetermineContractAddress` and `TrieIdGenerator` are no longer part of the `pallet_contracts::Trait`. They can just be removed from the configuration.

Additionally those types are removed from the pallet public interface:
```
 DefaultSignedClaimHandicap
 DefaultTombstoneDeposit
 DefaultStorageSizeOffset
 DefaultRentByteFee
 DefaultRentDepositOffset
 DefaultSurchargeReward
 DefaultMaxDepth
 DefaultMaxValueSize
```

The pallet contracts cannot possible define sane defaults for rent parameters because that is an economic decision each runtime has to make. The rest of the defaults is removed for consistency and similar reason. Have a look at the substrate node for guidance.

### CODE REVIEW

**Comment by seanyoung** [comment]
This is a good idea, and it will fit nicely into solidity where the salt can be specified on contract creation:
```
Piffle p = new Piffle{salt: 102}(arg1, arg2);
```
Currently Solang appends the salt to the input.

However, I do have concerns about updating the host function `seal_instantiate`. The various changes so far like renaming `ext_*` to `seal_*` and the metadata changes where painful for users.  This means again that the version of Solang and contracts module/plasm must line up; this is a breaking change. We could have `seal_instantiate_with_salt` rather than update the prototype of the existing host function.

**Comment by athei** [comment]
I understand your concerns. However, is is too early to start with compatibility. I am doing all the breaking changes right now. We start with semver and compatiblity once pallet_contracts is officially released/deployed. I know  it is "released" but only it is because it is part of the mono repo. In practice it is still considered pre-release.

**Comment by atenjin** [comment]
why not just like Ethereum, add `nonce ` as input to generate contracts address?

We always thought "hash(deploying_address ++ code_hash ++ hash(input_bytes))" may be your particular design...

So in our Patract RedSpot, we provide a tools to generate a accounts in repeated test to avoid this situation( what means we change deploying_address in every deploy)
refer:
https://polkadot.polkassembly.io/post/156
微信图片_20201110172656

And on the other hand, in our contract testnet Jupiter, we change the rule of generate an address like:

https://github.com/patractlabs/jupiter/blob/207bc7dc43398e59fed19ef2ad890bf6b11d4675/frame/contracts-ext/src/lib.rs#L73-L99

we add a stroage to distinguish whether this deploy need to use `nonce` as part of the rule of generating an address, so that in test or other situation would deploy normally.

We think the old design "hash(deploying_address ++ code_hash ++ hash(input_bytes))" may be useful in some situations. Add a `salt` changes the old default rule, I think if need to change address every, why not use `nonce`?

All in all, I think the old design could be reserved in some situation, so that the contracts could be unique in chain.

**Comment by athei** [comment]
> why not just like Ethereum, add nonce as input to generate contracts address?

Explained in the opening post: The new design is strictly more powerful. You are free to supply the account nonce as `salt` if you like. Just keep in mind that the `nonce` is only bumped when sending a transaction.

> We always thought "hash(deploying_address ++ code_hash ++ hash(input_bytes))" may be your particular design...

It was. And it is changing as explained in the opening post. For the why. I think you answered that yourself pretty well: You had to a a lot of gymnastics so that your tool works with the old semantic.

> We think the old design "hash(deploying_address ++ code_hash ++ hash(input_bytes))" may be useful in some situations. Add a salt changes the old default rule, I think if need to change address every, why not use nonce?
> 
> All in all, I think the old design could be reserved in some situation, so that the contracts could be unique in chain.

You can replicate the old behaviour yourself if you think that is useful to your use case. Just hash your input data and supply it as `salt`.

File: frame/contracts/src/lib.rs [hunk 1]
```diff
@@ -674,9 +614,34 @@ impl<T: Trait> Module<T> {
 		let result = wasm::save_code_raw::<T>(code, &schedule);
 		result.map(|_| ()).map_err(Into::into)
 	}
+
+	/// Determine the address of a contract,
+	///
+	/// This is the address generation function used by contract instantation. Its result
+	/// is only dependend on its inputs. It can therefore be used to reliably predict the
+	/// address of a contract. This is akin to the formular of eth's CRATE2 opcode. There
+	/// is no CREATE equivalent because CREATE2 is strictly more powerful.
```
Comment (NikVolf) [inline frame/contracts/src/lib.rs:623]: Sure we want ethereum terminology (CREATE2) in docs?

Comment (athei) [reply frame/contracts/src/lib.rs:623]: Did not think that intensely about that. What is your recommendation?

**Review by NikVolf — APPROVED** [review]
LGTM

**Review by wheresaddie — APPROVED** [review]
thanks!

**Review by ascjones — APPROVED** [review]
LGTM

**Comment by athei** [comment]
bot merge

**Pull request merged**